File: tinylog-sample/src/main/java/com/androidwind/log/sample/MainActivity.java
package com.androidwind.log.sample;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.text.method.ScrollingMovementMethod;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.androidwind.log.TinyLogConfig;
import com.androidwind.log.TinyLog;
import com.androidwind.permission.OnPermission;
import com.androidwind.permission.Permission;
import com.androidwind.permission.TinyPermission;

import java.util.List;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private String mResult;
    private StringBuilder mStringBuilder;
    private int times;
    private TextView result;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        handlePermission();

        Button btn1 = findViewById(R.id.btn_click_once);
        btn1.setOnClickListener(this);
        Button btn2 = findViewById(R.id.btn_click_more);
        btn2.setOnClickListener(this);
        Button btn3 = findViewById(R.id.btn_clear);
        btn3.setOnClickListener(this);
        result = findViewById(R.id.tv_output);
        result.setMovementMethod(ScrollingMovementMethod.getInstance());

        TinyLog.config().setLogCallBack(mLogCallBack);
        mStringBuilder = new StringBuilder();
    }

    private void handlePermission() {
        //permission
        TinyPermission.start(this)
                .permission(Permission.READ_EXTERNAL_STORAGE, Permission.WRITE_EXTERNAL_STORAGE)
                .request(new OnPermission() {

                    @Override
                    public void hasPermission(List<String> granted, boolean isAll) {

                    }

                    @Override
                    public void noPermission(List<String> denied, boolean permanent) {
                        TinyLog.config()
                                .setEnable(BuildConfig.DEBUG).setWritable(false).setLogLevel(TinyLog.LOG_I);
                    }
                });
    }

    private void printLog() {
        TinyLog.v("this is tinylog");
        TinyLog.d("TAG", "this is tinylog");
        TinyLog.i("TAG", "this is tinylog", "arg1");
        TinyLog.w("TAG", "this is tinylog", "arg1", "arg2");
        try {
            String s = null;
            s.toString();
        } catch (NullPointerException e) {
            TinyLog.e("TAG", e, "printStackTrace");
        }
        try {
            Thread.sleep(100);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_click_once) {
            mStringBuilder.delete(0, mStringBuilder.length());
            result.setText(mStringBuilder.toString());
            printLog();
        } else if (v.getId() == R.id.btn_click_more) {
            mStringBuilder.delete(0, mStringBuilder.length());
            result.setText(mStringBuilder.toString());

            EditText etTimes = findViewById(R.id.et_times);
            if (TextUtils.isEmpty(etTimes.getText().toString())) {
                return;
            }
            times = Integer.valueOf(etTimes.getText().toString());
            new Thread(new Runnable() {
                @Override
                public void run() {
                    //test
                    for (int i = 0; i < times; i++) {
                        printLog();
                    }
                }
            }).start();
        } else if (v.getId() == R.id.btn_clear) {
            mStringBuilder.delete(0, mStringBuilder.length());
            result.setText(mStringBuilder.toString());
        }
    }

    private TinyLogConfig.LogCallBack mLogCallBack = new TinyLogConfig.LogCallBack() {
        @Override
        public void getLogString(String tag, String content) {
            mStringBuilder.append(tag + " -----------> " + content).append("\n");
            MainActivity.this.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    result.setText(mStringBuilder.toString());
                }
            });
        }
    };
}

File: tinylog-sample-kotlin/src/main/java/com/androidwind/log/sample/MainActivity.kt
package com.androidwind.log.sample

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.text.method.ScrollingMovementMethod
import android.view.View

import com.androidwind.log.TinyLogConfig
import com.androidwind.log.TinyLog
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private val mResult: String? = null
    private val mStringBuilder: StringBuilder by lazy { StringBuilder() }
    private var times: Int = 0

    private val mLogCallBack = object : TinyLogConfig.LogCallBack {
        override fun getLogString(tag: String, content: String) {
            mStringBuilder.append("$tag -----------> $content").append("\n")
            this@MainActivity.runOnUiThread { tv_output?.text = mStringBuilder.toString() }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btn_click_once.setOnClickListener(this)
        btn_click_more.setOnClickListener(this)
        btn_clear.setOnClickListener(this)
        tv_output?.movementMethod = ScrollingMovementMethod.getInstance()

        TinyLog.config().setLogCallBack(mLogCallBack)
    }

    private fun printLog() {
        TinyLog.v("this is tinylog")
        TinyLog.d("TAG", "this is tinylog")
        TinyLog.i("TAG", "this is tinylog", "arg1")
        TinyLog.w("TAG", "this is tinylog", "arg1", "arg2")
        try {
            val s: String? = null
            s!!.toString()
        } catch (e: NullPointerException) {
            TinyLog.e("TAG", e, "printStackTrace")
        }

        try {
            Thread.sleep(100)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    override fun onClick(v: View) {
        if (v.id == R.id.btn_click_once) {
            mStringBuilder.delete(0, mStringBuilder.length)
            tv_output?.text = mStringBuilder.toString()
            printLog()
        } else if (v.id == R.id.btn_click_more) {
            mStringBuilder.delete(0, mStringBuilder.length)
            tv_output?.text = mStringBuilder.toString()
            if (TextUtils.isEmpty(et_times.getText().toString())) {
                return
            }
            times = Integer.valueOf(et_times.getText().toString())
            Thread(Runnable {
                //test
                for (i in 0 until times) {
                    printLog()
                }
            }).start()
        } else if (v.id == R.id.btn_clear) {
            mStringBuilder.delete(0, mStringBuilder.length)
            tv_output?.text = mStringBuilder.toString()
        }
    }
}
